cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -188,38 +208,59 @@ runQueryTip (AnyConsensusModeParams cModeParams) network mOutFile = do
         ChainTipAtGenesis -> 0
         ChainTip slotNo _ _ -> slotNo
 
-  mEpoch <- mSlotToEpoch consensusMode localNodeConnInfo tipSlotNo
-    & fmap tuple3Fst
-    & logExceptContinue renderShelleyQueryCmdError
+  case consensusMode of
+    CardanoMode -> do
+      mEraHistoryResult :: Maybe (EpochNo, Either Aeson.Value RelativeTime) <- logExceptContinue renderShelleyQueryCmdError $ do
+        eResult <- liftIO $ queryNodeLocalState localNodeConnInfo Nothing (QueryEraHistory CardanoModeIsMultiEra)
+        case eResult of
+          Left acqFail -> left (ShelleyQueryCmdAcquireFailure acqFail)
+          Right eraHistory -> do
+            a <- case slotToEpoch tipSlotNo eraHistory of
+              Left e -> throwE (ShelleyQueryCmdPastHorizon e)
+              Right a -> return (tuple3Fst @EpochNo @SlotsInEpoch @SlotsToEpochEnd a)
+
+            let b = first toJsonPastHorizonException $ fmap fst (getProgress tipSlotNo eraHistory)
 
-  let output = encodePretty
-        . toObject "era" (Just (toJSON anyEra))
-        . toObject "epoch" (Just mEpoch)
-        $ toJSON tip
-  case mOutFile of
-    Just (OutputFile fpath) -> liftIO $ LBS.writeFile fpath output
-    Nothing                 -> liftIO $ LBS.putStrLn        output
+            return (a, b)
 
-  where
-    tuple3Fst :: (a, b, c) -> a
-    tuple3Fst (a, _, _) = a
+      let mEpoch = fst <$> mEraHistoryResult
+      let tipTimeResult = case mEraHistoryResult of
+            Nothing -> Left $ Aeson.object
+              [ "error" .= Aeson.String "tip time missing"
+              ]
+            Just a -> snd a
 
-    mSlotToEpoch
-      :: ConsensusMode mode
-      -> LocalNodeConnectInfo mode
-      -> SlotNo
-      -> ExceptT ShelleyQueryCmdError IO (EpochNo, SlotsInEpoch, SlotsToEpochEnd)
-    mSlotToEpoch cMode lNodeConnInfo slotNo = case cMode of
-      CardanoMode -> do
-        let epochQuery = QueryEraHistory CardanoModeIsMultiEra
-        eResult <- liftIO $ queryNodeLocalState lNodeConnInfo Nothing epochQuery
+      systemStart <- do
+        eResult <- liftIO $ queryNodeLocalState localNodeConnInfo Nothing QuerySystemStart
         case eResult of
-          Left acqFail -> left (ShelleyQueryCmdAcquireFailure acqFail)
-          Right eraHistory -> case slotToEpoch slotNo eraHistory of
-            Left e -> throwE (ShelleyQueryCmdPastHorizon e)
-            Right a -> return a
+          Left acqFail -> left (ShelleyQueryCmdGeneric (show acqFail))
```

Comment:
We already have `ShelleyQueryCmdAcquireFailure` for the `AcquireFailure` type.

Reply:
Fixed!